soco/zonegroupstate.py

```diff
@@ -0,0 +1,267 @@
+        self.processed_count = 0
+
+    def clear_cache(self):
+        """Clear the cache timestamp."""
+        self._cache_until = NEVER_TIME
+
+    def clear_zone_groups(self):
+        """Clear all known group sets."""
+        self.groups.clear()
+        self.all_zones.clear()
+        self.visible_zones.clear()
+
+    def poll(self, soco):
+        """Poll using the provided SoCo instance and process the payload."""
+        # pylint: disable=protected-access
+        if time.monotonic() < self._cache_until:
+            self.total_requests += 1
+            _LOG.debug(
+                "Cache still active (GetZoneGroupState) during poll for %s",
+                soco.ip_address,
+            )
+            return
+
+        if soco._is_satellite:
+            # Satellites can return outdated information, use the parent
+            _LOG.debug(
+                "Poll request on satellite (%s), using parent (%s)",
+                soco.ip_address,
+                soco._satellite_parent.ip_address,
+            )
+            soco = soco._satellite_parent
+
+        zgs = soco.zoneGroupTopology.GetZoneGroupState()["ZoneGroupState"]
+        self.process_payload(zgs, "poll", soco.ip_address)
+
+    def process_payload(self, payload, source, source_ip):
+        """Update using the provided XML payload."""
+        self.total_requests += 1
+
+        def update_cache():
+            if source == "event":
+                timeout = EVENT_CACHE_TIMEOUT
+            else:
+                timeout = POLLING_CACHE_TIMEOUT
+            self._cache_until = time.monotonic() + timeout
+            _LOG.debug("Setting ZGS cache to %ss", timeout)
+
+        tree = normalize_zgs_xml(payload)
+        normalized_zgs = str(tree)
+        if normalized_zgs == self._last_zgs:
+            _LOG.debug(
+                "Duplicate ZGS received from %s (%s), ignoring", source_ip, source
+            )
+            update_cache()
+            return
+
+        self.processed_count += 1
+        _LOG.debug(
+            "Updating ZGS with %s payload from %s (%s/%s processed)",
+            source,
+            source_ip,
+            self.processed_count,
+            self.total_requests,
+        )
+
+        self.update_soco_instances(tree)
+        update_cache()
+        self._last_zgs = normalized_zgs
+
+    def parse_zone_group_member(self, member_element):
+        """Parse a ZoneGroupMember or Satellite element from Zone Group
+        State, create a SoCo instance for the member, set basic attributes
+        and return it."""
+        # pylint: disable=protected-access
+
+        # Create a SoCo instance for each member. Because SoCo
+        # instances are singletons, this is cheap if they have already
+        # been created, and useful if they haven't. We can then
+        # update various properties for that instance.
+        member_attribs = member_element.attrib
+        ip_addr = member_attribs["Location"].split("//")[1].split(":")[0]
```

Comment:
For extraction of information from strings I find it tremendously helpful with an example string in a comment above right above it. I know that it is mentioned in dosctring at the module level, but I find that the locality makes it easier to read.

Reply:
This was copied over from `core.py` as-is. I'll add a comment with an example.

Reply:
Done.